src/View/AjaxView.php

```diff
@@ -95,39 +103,66 @@ public function render($view = null, $layout = null) {
 		}
 
 		if ($view !== false && !isset($this->viewVars['_redirect']) && $this->_getViewFileName($view)) {
-			$response['content'] = parent::render($view, $layout);
+			$dataToSerialize['content'] = parent::render($view, $layout);
 		}
+
+		$this->viewVars = Hash::merge($dataToSerialize, $this->viewVars);
+
 		if (isset($this->viewVars['_serialize'])) {
-			$response = $this->_serialize($response, $this->viewVars['_serialize']);
+			$dataToSerialize = $this->_dataToSerialize($this->viewVars['_serialize'], $dataToSerialize);
 		}
-		$result = json_encode($response);
+
+		return $this->_serialize($dataToSerialize);
+	}
+
+	/**
+	 * Serialize(json_encode) accumulated data from both our custom render method
+	 *   and viewVars set by the user.
+	 *
+	 * @param array $dataToSerialize Array of data that is to be serialzed.
+	 * @return array The serialized data.
+	 */
+	protected function _serialize($dataToSerialize = []) {
+		$result = json_encode($dataToSerialize);
 		if (json_last_error() !== JSON_ERROR_NONE) {
 			return json_encode(['error' => json_last_error_msg()]);
 		}
 		return $result;
 	}
 
 	/**
-	 * Serializes view vars.
+	 * Returns data to be serialized based on the value of viewVars.
 	 *
-	 * @param array $response Response data array.
-	 * @param array $serialize The viewVars that need to be serialized.
-	 * @return array The serialized data.
+	 * @param array|string|bool $serialize The name(s) of the view variable(s) that
+	 *   need(s) to be serialized. If true all available view variables will be used.
+	 * @param array $additionalData Data items that were defined internally in our own
+	 *   render method.
+	 * @return mixed The data to serialize.
 	 */
-	protected function _serialize($response, $serialize) {
-		if (is_array($serialize)) {
-			foreach ($serialize as $alias => $key) {
-				if (is_numeric($alias)) {
-					$alias = $key;
-				}
-				if (array_key_exists($key, $this->viewVars)) {
-					$response[$alias] = $this->viewVars[$key];
-				}
+	protected function _dataToSerialize($serialize, $additionalData = []) {
+		if ($serialize === true) {
+			$data = array_diff_key(
+				$this->viewVars,
+				array_flip($this->_specialVars)
+			);
+
+			if (empty($data)) {
+				return null;
```

Comment:
Why is this needed? Shouldnt empty data be OK?

Reply:
Yes I believe you are probably right. I will test it and make the appropriate changes. Again the majority of this method came from the JsonView so I just assumed they knew what they were doing lol.

Reply:
OK
otherwise you need to adjust the doc block for

    protected function _serialize($dataToSerialize = []) {

to also allow null :)

Reply:
The chances this conditional is met are very slim, but I guess the answer to whether it should be removed or kept comes down to what the default json_encoded ajax value should if there are indeed no viewVars. In its current state, returning null, the default ajax value is null. If we remove the conditional the default ajax value is an empty array. Which do you prefer? I guess the more I think about it the condition can never really be true because we push data into viewVars in the render method...

Reply:
Exactly, so that case just cant happen, and array always is fine IMO.

Reply:
Cool I'll get rid of it right now.